// apollo-client/core/LocalState.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
import { print, } from 'graphql';
import graphql from 'graphql-anywhere';
import { mergeDeep, getMainDefinition, buildQueryFromSelectionSet, hasDirectives, checkDocument, removeDirectivesFromDocument, } from 'apollo-utilities';
import { capitalizeFirstLetter } from '../util/capitalizeFirstLetter';
var LocalState = (function () {
    function LocalState(_a) {
        var cache = _a.cache, client = _a.client, initializers = _a.initializers, resolvers = _a.resolvers, typeDefs = _a.typeDefs, fragmentMatcher = _a.fragmentMatcher;
        this.resolvers = {};
        this.firedInitializers = [];
        this.cache = cache;
        if (client) {
            this.client = client;
        }
        if (initializers) {
            this.runInitializersSync(initializers);
        }
        if (resolvers) {
            this.addResolvers(resolvers);
        }
        if (typeDefs) {
            this.setTypeDefs(typeDefs);
        }
        if (fragmentMatcher) {
            this.setFragmentMatcher(fragmentMatcher);
        }
    }
    LocalState.prototype.runInitializers = function (initializers) {
        var _this = this;
        if (!initializers) {
            throw new Error('Invalid/missing initializers');
        }
        var mergedInitializers = this.mergeInitializers(initializers);
        var initializerPromises = [];
        this.processInitializers(mergedInitializers, function (fieldName, initializer) {
            initializerPromises.push(Promise.resolve(initializer()).then(function (result) {
                var _a;
                if (result !== null) {
                    _this.cache.writeData({ data: (_a = {}, _a[fieldName] = result, _a) });
                }
            }));
        });
        return Promise.all(initializerPromises);
    };
    LocalState.prototype.runInitializersSync = function (initializers) {
        var _this = this;
        if (!initializers) {
            throw new Error('Invalid/missing initializers');
        }
        var mergedInitializers = this.mergeInitializers(initializers);
        this.processInitializers(mergedInitializers, function (fieldName, initializer) {
            var _a;
            var result = initializer(_this);
            if (result !== null) {
                _this.cache.writeData({ data: (_a = {}, _a[fieldName] = result, _a) });
            }
        });
    };
    LocalState.prototype.addResolvers = function (resolvers) {
        var _this = this;
        if (Array.isArray(resolvers)) {
            resolvers.forEach(function (resolverGroup) {
                _this.resolvers = mergeDeep(_this.resolvers, resolverGroup);
            });
        }
        else {
            this.resolvers = mergeDeep(this.resolvers, resolvers);
        }
    };
    LocalState.prototype.setResolvers = function (resolvers) {
        this.resolvers = {};
        this.addResolvers(resolvers);
    };
    LocalState.prototype.getResolvers = function () {
        return this.resolvers;
    };
    LocalState.prototype.runResolvers = function (_a) {
        var query = _a.query, remoteResult = _a.remoteResult, context = _a.context, variables = _a.variables, onError = _a.onError;
        var localResult = {};
        if (query) {
            var resolver = this.prepareResolver(query).resolver;
            if (resolver) {
                var rootValue = this.buildRootValueFromCache(query, variables);
                rootValue = rootValue
                    ? mergeDeep(rootValue, remoteResult)
                    : remoteResult;
                try {
                    localResult = graphql(resolver, query, rootValue, context, variables, { fragmentMatcher: this.fragmentMatcher });
                }
                catch (error) {
                    if (onError) {
                        onError(error);
                        return;
                    }
                }
            }
        }
        return __assign({}, remoteResult, localResult);
    };
    LocalState.prototype.setTypeDefs = function (typeDefs) {
        this.typeDefs = typeDefs;
    };
    LocalState.prototype.getTypeDefs = function () {
        return this.typeDefs;
    };
    LocalState.prototype.setFragmentMatcher = function (fragmentMatcher) {
        this.fragmentMatcher = fragmentMatcher;
    };
    LocalState.prototype.getFragmentMatcher = function () {
        return this.fragmentMatcher;
    };
    LocalState.prototype.clientQuery = function (document) {
        return hasDirectives(['client'], document) ? document : null;
    };
    LocalState.prototype.serverQuery = function (document) {
        return this.removeClientSetsFromDocument(document);
    };
    LocalState.prototype.prepareContext = function (context) {
        if (context === void 0) { context = {}; }
        var cache = this.cache;
        var schemas = [];
        if (this.typeDefs) {
            var directives = 'directive @client on FIELD';
            var definition = this.normalizeTypeDefs(this.typeDefs);
            schemas = schemas.concat([{ definition: definition, directives: directives }]);
        }
        var newContext = __assign({}, context, { cache: cache, getCacheKey: function (obj) {
                if (cache.config) {
                    return cache.config.dataIdFromObject(obj);
                }
                else {
                    throw new Error('To use context.getCacheKey, you need to use a cache that has ' +
                        'a configurable dataIdFromObject, like apollo-cache-inmemory.');
                }
            }, schemas: schemas });
        return newContext;
    };
    LocalState.prototype.addExportedVariables = function (document, variables, context) {
        if (variables === void 0) { variables = {}; }
        if (context === void 0) { context = {}; }
        var exportedVariables = {};
        if (document &&
            hasDirectives(['client'], document) &&
            hasDirectives(['export'], document)) {
            var preparedResolver = this.prepareResolver(document);
            if (preparedResolver.resolver) {
                var rootValue = this.buildRootValueFromCache(document, variables);
                var updatedContext = this.prepareContext(context);
                graphql(preparedResolver.resolver, document, rootValue || {}, updatedContext, variables);
                exportedVariables = preparedResolver.exportedVariables;
            }
        }
        return __assign({}, variables, exportedVariables);
    };
    LocalState.prototype.resetInitializers = function () {
        this.firedInitializers = [];
    };
    LocalState.prototype.mergeInitializers = function (initializers) {
        var mergedInitializers = {};
        if (Array.isArray(initializers)) {
            initializers.forEach(function (initializerGroup) {
                mergedInitializers = __assign({}, mergedInitializers, initializerGroup);
            });
        }
        else {
            mergedInitializers = initializers;
        }
        return mergedInitializers;
    };
    LocalState.prototype.processInitializers = function (initializers, runFunc) {
        var _this = this;
        Object.keys(initializers).forEach(function (fieldName) {
            if (_this.firedInitializers.indexOf(fieldName) < 0) {
                runFunc(fieldName, initializers[fieldName]);
                _this.firedInitializers.push(fieldName);
            }
        });
    };
    LocalState.prototype.prepareResolver = function (query) {
        var _this = this;
        var resolver = null;
        var exportedVariables = {};
        if (!query) {
            return {
                resolver: resolver,
                exportedVariables: exportedVariables,
            };
        }
        var definition = getMainDefinition(query);
        var definitionOperation = definition.operation;
        var type = definitionOperation
            ? capitalizeFirstLetter(definitionOperation)
            : 'Query';
        var _a = this, cache = _a.cache, client = _a.client;
        resolver = function (fieldName, rootValue, args, context, info) {
            if (rootValue === void 0) { rootValue = {}; }
            var resultKey = info.resultKey;
            var exportedVariable;
            if (info.directives && info.directives.export) {
                exportedVariable = info.directives.export.as;
            }
            var childRootValue = {};
            var rootValueKey = Object.keys(rootValue)[0];
            if (rootValueKey) {
                childRootValue = rootValue[rootValueKey];
            }
            var normalNode = rootValue[fieldName] !== undefined
                ? rootValue[fieldName]
                : childRootValue[fieldName];
            var aliasedNode = rootValue[resultKey] !== undefined
                ? rootValue[resultKey]
                : childRootValue[resultKey];
            var aliasUsed = resultKey !== fieldName;
            var field = aliasUsed ? fieldName : resultKey;
            var updatedContext = __assign({}, context, { cache: cache,
                client: client });
            var result;
            var resolverType = rootValue.__typename || type;
            var resolverMap = _this.resolvers[resolverType];
            if (resolverMap) {
                var resolve = resolverMap[field];
                if (resolve) {
                    result = resolve(rootValue, args, updatedContext, info);
                }
            }
            if (!result) {
                if (normalNode !== undefined || aliasedNode !== undefined) {
                    result = aliasedNode || normalNode;
                }
            }
            if (exportedVariable) {
                exportedVariables[exportedVariable] = result;
            }
            return result;
        };
        return {
            resolver: resolver,
            exportedVariables: exportedVariables,
        };
    };
    LocalState.prototype.buildRootValueFromCache = function (document, variables) {
        var query = buildQueryFromSelectionSet(document);
        var cachedData = this.cache.diff({
            query: query,
            variables: variables,
            optimistic: false,
        });
        return cachedData.result;
    };
    LocalState.prototype.removeClientSetsFromDocument = function (query) {
        checkDocument(query);
        var docClone = removeDirectivesFromDocument([
            {
                test: function (directive) { return directive.name.value === 'client'; },
                remove: true,
            },
        ], query);
        if (docClone) {
            var nonClientDefinitions_1 = [];
            docClone.definitions.forEach(function (definition) {
                var isTypenameOnly = definition.selectionSet.selections.every(function (selection) {
                    return (selection.kind === 'Field' &&
                        selection.name.value === '__typename');
                });
                if (!isTypenameOnly) {
                    nonClientDefinitions_1.push(definition);
                }
            });
            docClone.definitions = nonClientDefinitions_1;
        }
        return docClone;
    };
    LocalState.prototype.normalizeTypeDefs = function (typeDefs) {
        var defs = Array.isArray(typeDefs) ? typeDefs : [typeDefs];
        return defs
            .map(function (typeDef) { return (typeof typeDef === 'string' ? typeDef : print(typeDef)); })
            .map(function (str) { return str.trim(); })
            .join('\n');
    };
    return LocalState;
}());
export { LocalState };
